api/API_VERSIONING.md

```diff
@@ -28,11 +29,31 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
 v3 API bootstrap configuration is `envoy.config.bootstrap.v3.Boostrap`, even
 though it might might transitively reference `envoy.service.trace.v2`.
 
+The minor version enables API stability and support, and to safely deprecate old
+features and reduce the tech debt. A client can support a range of minor API
+versions, and can request xDS resources that match this range, using the resource
+[context parameters](https://github.com/cncf/udpa/blob/master/xds/core/v3/context_params.proto).
```

Comment:
I would still like to consider adding a flexible range-based matching semantic for minor versions.  Can we do that as part of this effort, or do we want to decouple that from this?

Hmm...  Actually, now that we've decided to stick with string representations for new-style resource names, doesn't that mean that we don't actually have a way to add more flexible matching syntax later if we don't do it now?  When we were going to be using `ResourceLocator` on the wire, it would have been easy to add new matching types, because they'd be expressed in different fields.  But with the string form, it's basically just "key=value&key=value&...".  So if we're interpreting the values literally, then any new special syntax we add there may break existing users.

It seems like we need to either add the more flexible matching syntax now or carve out some sort of escape hatch in the URI syntax to allow for future changes.  (Actually, the latter would probably be a good idea even if we do the former, since we may want to add other flexible matcher types in the future.)

Reply:
I'm on the fence on this one. I would like to future proof with some notion of escape hatch in the URI syntax, that's a no brainer. I think I'd like to understand the full set of flexible match objectives before deciding on a concrete syntax and what do use here instead of min/max version. How do you think is best to go forward on that @markdroth?

Reply:
I agree that range-based matching semantics parameters should be supported, but I'm not certain I understand many of the use-cases that it will be used for.
I think that eventually we can replace this with a parameter similar to this: `xds.api.range.minor_version={"start":5,"end":6}`.

Reply:
Is the concern here that we don't want magic params that have prior agree to semantics? Meaning, if we have an oldest value and a newest value it seems not hard to just do the range matching manually.

Reply:
IIUC the issue of exact-matching is that the returned resource must have the same parameters.
Range-parameters enable explicit semantics. For example, a cache (or xds-relay) will be able to return the same resource with minor version 5 for a request with the range [4,5] and for a different request with the range [5,6].

Reply:
Right, the issue is that the current exact-match semantic means that the context params in the result must exactly match the ones in the request.  So if you have one client requesting `xds.api.range.minor_version=[0, 1]` and another requesting `xds.api.range.minor_version=[1, 2]`, you can provide the contents at version 1 to both of them, but you need to return each one with a different context parameter.  This will basically lead to cache pollution, where any caching proxy will need an increasing number of copies of each identical resource under different names, and the pollution will increase combinatorically.

What I am proposing is a range-based matching mechanism where the clients can ask for something like `xds.api.range.minor_version=SPECIAL_RANGE_SYNTAX[0, 1]` and `xds.api.range.minor_version=SPECIAL_RANGE_SYNTAX[1, 2]`, and both can get back the same resource with context parameter `xds.api.range.minor_version=1`.  In order to make that work, we need to define the syntax for range-based matching, make sure it doesn't conflict with any "flat" context parameter values (e.g., by defining some sort of escaping mechanism), and make sure that xDS servers, clients, and proxies understand it.

This is a bit more work, but I think it's also a cleaner solution.  And I think this same mechanism will be able to be used for other use-cases in the future.  (We already have one use-case internally for which we've been discussing something like this.)

@htuch, assuming that we have a reasonable way to future-proof the syntax in a way that would allow us to add more flexible matching types if we need to do so in the future, I think I'd advocate for restricting our initial changes to just the cases that we need immediately.  For minor version, I think that's range-based matching.  And if we want to handle the patch version the same way, we might also want a simple wildcard match for that.  I think anything else can wait for the future.

I would be happy to write up a proposal for how this would work, but I need something to build off of.  Where did we leave things on plans for the initial xRFC for the new naming scheme?  Are you still planning to do that at some point?

Reply:
OK thanks this makes sense to me. I agree we should do the work to implement range syntax. It doesn't seem too terribly difficult.

Reply:
Yes, a proposal here for a syntax which is future proof and can initially accommodate simple range would help unblock work here I think. Thanks Mark.